src/UpgradeExecRouter.sol

```diff
@@ -0,0 +1,240 @@
+        onlyRole(DEFAULT_ADMIN_ROLE)
+    {
+        _setMinL1TimelockDelay(_minL1TimelockDelay);
+    }
+
+    function routeActions(
+        uint256[] memory chainIds,
+        address[] memory actionAddresses,
+        uint256[] memory actionDatavalues,
+        bytes[] memory actionData,
+        bytes32 timelockSalt
+    ) public view returns (address, bytes memory) {
+        require(chainIds.length == actionAddresses.length, "CoreProposalCreator: length mismatch");
+        require(chainIds.length == actionDatavalues.length, "CoreProposalCreator: length mismatch");
+        require(chainIds.length == actionData.length, "CoreProposalCreator: length mismatch");
+
+        address[] memory schedTargets = new address[](chainIds.length);
+        uint256[] memory schedValues = new uint256[](chainIds.length);
+        bytes[] memory schedData = new bytes[](chainIds.length);
+        //CHRIS: TODO: check arrays are same length
+        for (uint256 i = 0; i < chainIds.length; i++) {
+            require(
+                upExecLocations.exists(chainIds[i]),
+                "UpgradeExecRouter: Upgrade exec location does not exist"
+            );
+            UpExecLocation memory upExecLocation = upExecLocations[chainIds[i]];
+
+            bytes memory executorData = abi.encodeWithSelector(
+                UpgradeExecutor.execute.selector, actionAddresses[i], actionData[i]
+            );
+
+            // CHRIS: TODO: safety check that all targets and data are non zero?
+            if (upExecLocation.inbox == address(0)) {
+                schedTargets[i] = upExecLocation.upgradeExecutor;
+                schedValues[i] = actionDatavalues[i];
+                schedData[i] = executorData;
+            } else {
+                // For L2 actions, magic is top level target, and value and calldata are encoded in payload
+                schedTargets[i] = RETRYABLE_TICKET_MAGIC;
+                schedValues[i] = 0;
+                schedData[i] = abi.encode(
+                    upExecLocation.inbox,
+                    upExecLocation.upgradeExecutor,
+                    actionDatavalues[i],
+                    0,
+                    0,
+                    executorData
+                );
+            }
+        }
+
+        // schedule that
+        bytes memory timelockCallData = abi.encodeWithSelector(
+            L1ArbitrumTimelock.scheduleBatch.selector,
+            schedTargets,
+            schedValues,
+            schedData,
+            DEFAULT_PREDECESSOR, // CHRIS: TODO: should we allow others here? could be important? no, use the complex route?
+            timelockSalt,
+            l1TimelockMinDelay
+        );
+
+        return (
+            address(100),
+            abi.encodeWithSelector(ArbSys.sendTxToL1.selector, l1TimelockAddr, timelockCallData)
+        );
+    }
+
+    function routeActionsWithDefaults(
+        uint256[] memory chainIds,
+        address[] memory actionAddresses,
+        bytes32 timelockSalt
```

Comment:
for the withDefaults path, could the salt be calculated in this contract instead?

Reply:
yeh, i'm not sure what to do about that yet, will put a TODO and merge